frontend/src/component/Activity/Preview.jsx

```diff
@@ -0,0 +1,239 @@
+        content.workspace_id,
+        content.content_id,
+        content.current_revision_id,
+        filenameNoExtension,
+        FIRST_PAGE,
+        width,
+        MAX_PREVIEW_HEIGHT
+      )
+    )
+
+    const src = ([mediaQuery, width]) => `${prev(width)} ${width}w`
+
+    return (
+      <div class='activityFeed__preview__image'>
+        <img
+          alt={this.props.t('Preview of {{content}}', { content: content.label })}
+          title={content.label}
+          src={prev(PREVIEW_WIDTHS[0][1])} // fall back on the smallest image size
+          srcset={PREVIEW_WIDTHS.map(src).join(',')}
+          sizes={
+            PREVIEW_WIDTHS
+              .map(([mediaQuery, width]) => `${mediaQuery} ${width}px`.trim())
+              .join(',')
+          }
+          onError={() => this.handleUnavailablePreview()}
+        />
+      </div>
+    )
+  }
+
+  async updatePreview () {
+    let previewComponent = null
+
+    if (this.isHtmlPreview()) {
+      if (!this.state.previewComponent) {
+        this.setLoadingPreview()
+      }
+
+      try {
+        previewComponent = await this.getHTMLPreviewComponent()
+      } catch (e) {
+        this.handleUnavailablePreview()
+        console.error('Unable to produce the preview of content ', this.props.content, e)
+        return
+      }
+
+    } else {
+      previewComponent = this.getJPEGPreview()
+    }
+
+    this.setState({
+      previewComponent,
+      previewLoading: false
+    })
+  }
+
+  receivePreviewRef (ref) {
+    this.previewRef = ref
+    this.testPreviewOverflow()
+  }
+
+  componentDidUpdate (prevProps) {
+    if (prevProps.content === this.props.content) {
+      this.testPreviewOverflow()
+    } else if (prevProps.content.current_revision_id !== this.props.content.current_revision_id) {
+      this.updatePreview()
+    }
+  }
+
+  componentDidMount () {
+    if (this.isHtmlPreview()) {
+      this.updatePreview()
+    } else {
+      this.testPreviewOverflow()
+    }
+  }
+
+  testPreviewOverflow () {
+    const ref = this.previewRef
+    if (ref) {
+      const previewOverflow = ref.scrollHeight > ref.offsetHeight
+      if (this.state.previewOverflow !== previewOverflow) {
+        this.setState({ previewOverflow })
+      }
+    }
+  }
+
+  render () {
+    const { props, state } = this
+    const { content } = props
+
+    return (
+      <div
+        className={classnames(
+          'activityFeed__preview', {
+            activityFeed__preview__overflow: state.previewOverflow,
+            activityFeed__preview__unavailable: state.previewUnavailable,
+            activityFeed__preview__loading: state.previewLoading
+        })}
+        ref={(ref) => this.receivePreviewRef(ref)}
+      >
+        <Link to={PAGE.WORKSPACE.CONTENT(content.workspace_id, content.content_type, content.content_id)}>
+          {(
+            state.previewLoading
+              ? props.t('Preview loading...')
+              : (
+                  state.previewUnavailable
```

Comment:
Rather than having a state for `previewUnavailable`, wouldn't it be easier to set the previewComponent to this "unavailable preview" component when the situation arises?

Reply:
I need the previewUnavailable field for the class name. I originally chose to not store the preview unavailable rendering in previewComponent to avoid having two sources of truth, but I'm fine with setting previewComponent, so I did this.